### ISSUE TITLE
Understand BclDateTimeZone failures under Mono

### ISSUE THREAD

**Issue by GoogleCodeExporter** [nodatime/nodatime#97]
```
BclDateTimeZoneTest.AllZoneTransitions is disabled on Mono, but the reason it's 
disabled appears to no longer be correct.

Under Mono 2.10.9,  it fails with e.g.

36) Test Failure : 
NodaTime.Test.TimeZones.BclDateTimeZoneTest.AllZoneTransitions(Europe/Athens)
     Incorrect offset at 1952-06-30T22:00:00Z in interval EEST:[1952-06-30T22:00:00Z, 1952-11-01T21:00:00Z) +03 (+01)
  Expected: 02:00:00
  But was:  03:00:00

Where the instant at start-of-zone interval is returning +3 rather than the BCL 
+2.

We also have some failures at the end of the zone intervals, e.g.

46) Test Failure : 
NodaTime.Test.TimeZones.BclDateTimeZoneTest.AllZoneTransitions(Pacific/Noumea)
     Incorrect offset at 1978-02-26T11:59:59Z in interval NCST:[1977-12-03T13:00:00Z, 1978-02-26T12:00:00Z) +12 (+01)
  Expected: 11:00:00
  But was:  12:00:00


I'm pretty certain this _isn't_ what was documented:
// This test is effectively disabled on Mono as its time zone support is broken 
in the current
// stable release - see http://bugzilla.xamarin.com/show_bug.cgi?id=326 for 
example: even asking
// for the offset from UTC *for a UTC DateTime* can fail. We can restore these 
tests when we either
// find something to fix in Noda Time, or Mono is fixed.

I suppose this could be https://bugzilla.xamarin.com/show_bug.cgi?id=3333, but 
I'm not convinced.
```

Original issue reported on code.google.com by `malcolm.rowe` on 30 Jul 2012 at 8:59

**Comment by haf** [comment]
Time to revisit this one? Mono has fixed a lot of their bugs in the timezone handling.

**Comment by jskeet** [comment]
I'm not sure I'd say "currently spreads FUD" - it specifically says "at the time of this writing" and it's a user guide for a release which is over 3 years old. I don't think it's reasonable to try to keep every bit of *old* user guide updated with bug fixes.

I agree it's worth revisiting the issue though - I'll try running the tests under recent versions of Mono, although it gets tricky when I remember seeing different issues with TimeZone.Local depending on whether I was running on Windows, MacOSX or Xamarin...

**Comment by haf** [comment]
You can add the matrix to .travis, like it's written here: https://github.com/SuaveIO/suave/blob/master/.travis.yml#L4

**Comment by jskeet** [comment]
Yes, I can try that - but getting it under *every* variant of Mono is going to be a pain. I'll see how easy it is to get up and running, but it's all quite a lot of extra work for a platform that isn't as significant IMO (outside Xamarin) as it used to be...

**Comment by haf** [comment]
<img width="966" alt="screen shot 2017-04-03 at 17 35 56" src="https://cloud.githubusercontent.com/assets/193115/24617396/0cc300a4-1894-11e7-952f-01b6e805ccf2.png">

Latest – that is 4.8.0 at time of writing, and latest - 1 (4.6.x) is enough IMO.

**Comment by jskeet** [comment]
I wonder whether it might be worth delaying all of this until the NUnit test support for the dotnet SDK is done. Maybe the immediate actions should be:

- A little bit of manual investigation, just trying to run locally and possibly on Xamarin if I can get that working on my iPad again
- Updating the user guides for at least 2.0.x and unstable

To be honest, testing on Travis isn't going to be nearly as useful as testing on actual Xamarin devices (Android/iOS) as that's where I'd expect Mono to be running these days (as opposed to .NET Core).

PRs welcome, of course...

**Comment by haf** [comment]
We've finished [expecto for .Net core](https://github.com/haf/expecto/issues/125) – a modern testing framework – I'd be happy to help you port the tests to a modern test framework! :)

**Comment by jskeet** [comment]
No, I have **absolutely no desire** to move off NUnit, which does everything I need it to.

**Comment by haf** [comment]
:) Roger that. It was worth a shot.

**Comment by haf** [comment]
Back on track then; Travis testing is probably useful for mono on the server, because .Net core just became stable like two weeks ago (from a F# viewpoint).

**Comment by jskeet** [comment]
Hmm. I suspect the intersection of folks using F# on Mono with Noda Time is pretty small, to be honest. (If you start with the pool of .NET developers, each of those three subsets will reduce the number quite a lot.) It would certainly be nice to know how well it works, but it's lower down my list of priorities than quite a few other things.

**Comment by haf** [comment]
I'm not interested in that subset, but the all-mono-users set, which includes me. Mono users haven't had proper tools for .Net Core for a long time either (or it's been really annoying to get started).

**Comment by jskeet** [comment]
Okay, after some jiggling around I've now got the net45/net451 build running on Mono 4.8.0 on Ubuntu.

It currently fails in one test, with this stack trace start:

```text
1) Error : NodaTime.Test.TimeZones.BclDateTimeZoneSourceTest.ProviderReturnsBclDateTimeZoneForAllAdvertisedIds
System.InvalidOperationException : Requested transition for a ZoneYearOffset of February 29th in a non-leap year, not moving backwards to find a day-of-week
  at NodaTime.TimeZones.ZoneYearOffset.GetOccurrenceForYear (System.Int32 year) [0x0006d] in <c90fbbee5fa9426392c49904a3d9b7d0>:0 
```

That's a pretty strange error - will need to look into it. I haven't started re-enabling the "deliberately skipped on Mono" tests, nor have I tried on Windows.

**Comment by jskeet** [comment]
The tests which run by default on Mono now pass... but re-enabling the TimeZoneInfo-based ones causes a lot of failures, at least on Linux. (Will try doing the same thing on Windows.)

I definitely still want to understand those... it gets complicated because Noda Time tries to emluate the BCL as closely as it can bar a couple of places where the BCL is clearly very broken. It's possible that Mono is treating the data as documented rather than as it's used in reality.

**Comment by jskeet** [comment]
Next bug on Linux is due to https://bugzilla.xamarin.com/show_bug.cgi?id=54480

**Comment by jskeet** [comment]
(I can try to work around this in tests by checking that Noda Time is doing the right thing by not using the actual system time zone, but the recreated one... aside from Malta, which will fail when recreated. Sigh.)

**Comment by jskeet** [comment]
Still doesn't fix everything - there's more brokenness here... will write it up and see whether it's just Mono...

**Comment by jskeet** [comment]
And the next bug... https://bugzilla.xamarin.com/show_bug.cgi?id=54488
(That's just one of the failures. Will see if there's a pattern here.)

**Comment by jskeet** [comment]
I think everything that's left after #750 is merged (if we merge it) is due to Mono bugs which have been filed.

If I can work out a way to predict which zones will be affected by the Mono bugs, I might close this bug, and file a new one to periodically re-evaluate with new versions of Mono.

**Comment by haf** [comment]
The remaining 26 failures; where are those? Could you add mono build notes to the readme and perhaps create an issue with those failures?

My aim would be to categorise them as minor/major and to see how they can be fixed.

**Comment by jskeet** [comment]
I think it's down to 14 failures now (I'd have to check) but they're basically due to Mono bugs. I've changed the documentation around Mono support accordingly (in the unstable and 2.0.x user guides). Basically `BclDateTimeZone` doesn't give the same answer on Mono as the original `TimeZoneInfo` does. In most cases it gives a *better* answer, admittedly. I'd rather not create another issue for that, as *this* is the issue for working out why Mono doesn't work.

I have a change we could merge that would test all time zones on Mono on Linux, displaying the break - but all the tests would still pass on .NET Core. We could potentially add categories for those tests so that if we get CI testing Mono, we could explicitly skip them...

The bug fix for #743 needs to go into a 2.0.1 build, as otherwise `BclDateTimeZone` will simply crash for some time zones... but at that point, it's fairly reasonable to use Mono and just be aware of that difference. If you don't use `BclDateTimeZone`, using the IANA time zones instead, I expect it to be reasonable already.

**Comment by jskeet** [comment]
Note: all tests pass on Mono 4.8.0 on Windows, although one is skipped (AwkwardLeapYears) as I expect that to be broken on Windows too.

**Comment by jskeet** [comment]
Mono 5.0 is now out... I'll give that a try when I'm back near a Linux machine.

**Comment by andreysavihin** [comment]
Im using Mono (JIT compiler version 5.16.0.220 (tarball Mon Nov 26 17:22:20 UTC 2018)), NodaTime.dll (Latest stable binaries: NodaTime-2.4.4.zip from http://nodatime.org/) and simple C# code file Program.cs:

```
using System;
using NodaTime.TimeZones;

namespace Test
{
    internal class Program
    {
        private static void Main(string[] args)
        {
            foreach (var tz in TimeZoneInfo.GetSystemTimeZones())
            {
                try
                {
                    BclDateTimeZone.FromTimeZoneInfo(tz);
                }
                catch (Exception e)
                {
                    Console.WriteLine(tz.Id);
                    Console.WriteLine(e.Message);
                }
            }
        }
    }
}
```

Then I try compile and run exe file:

```
# mcs /reference:NodaTime.dll -out:Test.exe Program.cs
# mono Test.exe
```

Method FromTimeZoneInfo throw System.ArgumentOutOfRangeException: Year, Month, and Day parameters describe an un-representable DateTime.
in next timezones  Europe/Vatican, Europe/San_Marino, Europe/Rome, Europe/Malta.

**Comment by jskeet** [comment]
Could you include the full stack trace? I'll try to reproduce this myself of course... I can't immediately tell whether this is the same one I saw earlier.

**Comment by andreysavihin** [comment]
mono Test.exe
Europe/Malta
Year, Month, and Day parameters describe an un-representable DateTime.
  at System.DateTime.DateToTicks (System.Int32 year, System.Int32 month, System.Int32 day) [0x00077] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0
  at System.DateTime..ctor (System.Int32 year, System.Int32 month, System.Int32 day) [0x00000] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0
  at System.TimeZoneInfo.TransitionPoint (System.TimeZoneInfo+TransitionTime transition, System.Int32 year) [0x00018] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0
  at System.TimeZoneInfo.TryGetTransitionOffset (System.DateTime dateTime, System.TimeSpan& offset, System.Boolean& isDst) [0x0008c] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0
  at System.TimeZoneInfo.GetUtcOffsetHelper (System.DateTime dateTime, System.TimeZoneInfo tz, System.Boolean& isDST) [0x00029] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0
  at System.TimeZoneInfo.GetUtcOffset (System.DateTime dateTime, System.Boolean& isDST) [0x00025] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0
  at System.TimeZoneInfo.GetUtcOffset (System.DateTime dateTime) [0x00000] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0
  at NodaTime.TimeZones.BclDateTimeZone+BclAdjustmentRule..ctor (System.TimeZoneInfo zone, System.TimeZoneInfo+AdjustmentRule rule) [0x0000d] in <acf94f9b2516433e8c71e50e386f6a7b>:0
  at NodaTime.TimeZones.BclDateTimeZone+<>c__DisplayClass9_0.<FromTimeZoneInfo>b__0 (System.TimeZoneInfo+AdjustmentRule rule) [0x00000] in <acf94f9b2516433e8c71e50e386f6a7b>:0
  at System.Array.ConvertAll[TInput,TOutput] (TInput[] array, System.Converter`2[TInput,TOutput] converter) [0x00029] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0
  at NodaTime.TimeZones.BclDateTimeZone.FromTimeZoneInfo (System.TimeZoneInfo bclZone) [0x00083] in <acf94f9b2516433e8c71e50e386f6a7b>:0
  at Test.Program.Main (System.String[] args) [0x00017] in <592ffa47177449429320277a8d9cd3ad>:0

**Comment by jskeet** [comment]
Just to say, I've got as far as reproducing the problem. Will investigate further over the weekend.

**Comment by jskeet** [comment]
Right, I've just confirmed, this is the same bug as before, at least for Malta - the rule has a `DaylightTransitionStart` for February 29th.
While we could go to extravagant lengths to work around this, I don't think we should.

**Comment by jskeet** [comment]
Closing this, as it's now quite hard to run the tests on Mono, and Mono is increasingly less relevant anyway.
